src/Events/lib.mo

```diff
@@ -81,43 +80,90 @@ module {
+        canisters : [Principal];
+        allowType : HolderAllowType;
+    };
+
+    public type HolderAllowType = {
+        // Indicates that any NFT provides you unlimited spots.
+        #Unlimited;
     };
 
     public type Access_ = {
         #Public;
         #Private : Allowlist_;
+        #Holders : HolderAccess;
     };
 
     public module Access = {
         public func toStable(access_ : Access_) : Access = switch (access_) {
-            case (#Public)        #Public;
-            case (#Private(list)) #Private(Allowlist.toStable(list));
+            case (#Public)          #Public;
+            case (#Private(list))   #Private(Allowlist.toStable(list));
+            case (#Holders(access)) #Holders(access);
         };
 
         public func fromStable(access : Access) : Access_ = switch (access) {
-            case (#Public)        #Public;
-            case (#Private(list)) #Private(Allowlist.fromStable(list));
+            case (#Public)          #Public;
+            case (#Private(list))   #Private(Allowlist.fromStable(list));
+            case (#Holders(access)) #Holders(access);
         };
 
-        public func getSpots(access_ : Access_, user : Principal) : Int = switch (access_) {
-            case (#Public)        -1;
-            case (#Private(list)) switch (list.get(user)) {
-                case (null)    0;
-                case (? spots) switch (spots) {
+        public func getSpots(access_ : Access_, user : Principal) : async Int {
+            switch (access_) {
+                case (#Public)        -1;
+                case (#Private(list)) switch (list.get(user)) {
                     case (null)    0;
-                    case (? spots) spots;
+                    case (? spots) switch (spots) {
+                        case (null)    0;
+                        case (? spots) spots;
+                    };
+                };
+                case (#Holders(access)) switch (access.allowType) {
+                    case (#Unlimited) {
+                        for (cId in access.canisters.vals()) {
+                            let c : BalanceInterface = actor(Principal.toText(cId));
+                            let amount = try (await c.launchpadBalanceOf(user)) catch(_) { 0 };
+                            if (amount != 0) return -1;
+                        };
+                        0; // No matches.
+                    };
                 };
             };
         };
 
-        public func removeSpot(access_ : Access_, user : Principal) : Result<Int> = switch (access_) {
-            case (#Public)        #ok(-1);
-            case (#Private(list)) Allowlist.removeSpot(list, user);
+        public func removeSpot(access_ : Access_, user : Principal) : async Result<()> {
+            switch (access_) {
+                case (#Public)        #ok;
+                case (#Private(list)) Allowlist.removeSpot(list, user);
+                case (#Holders(access)) switch (access.allowType) {
```

Comment:
It's not clear to me why we need remove spots logic. Holders events are only unlimited, ♾ - 1 = ♾

Reply:
It looks like this func might be used as a guard? I.e. it is a hasSpots as well as a removeSpots?

Reply:
Yes, it basically does both, since you need to have spots to be able to remove it.